### ISSUE TITLE
Support for using functions in filters

### ISSUE THREAD

**Issue by kalombos** [Infinidat/infi.clickhouse_orm#73]
I think it would be great to have an opportunity to use functions in queries. For example:

```
select count(*) from storageunithistory where pointInPolygon((lat, lon), ...)
```
Is there a workaround for this?

**Comment by ishirav** [comment]
Functions would be a nice addition, but the Python syntax would be tricky... 
We can't do this:
`queryset.filter(pointInPolygon(...)=1)`
so maybe you'll need to define a name for the function's result first, like this:
`queryset.define(p=pointInPolygon(...)).filter(p=1)`
What do you think? Do you have a better idea?

**Comment by emakarov** [comment]
maybe something like
```
queryset.filter(F('functionName', *args))
```
?

**Comment by ishirav** [comment]
The syntax you suggested is suitable for boolean functions, but not when the user wants to compare the function's value to something else.
Also, I don't see the added value of doing
`queryset.filter(F('functionName', *args))`
and not simply passing a string:
`queryset.filter('functionName(x, y, z)')`

**Comment by emakarov** [comment]
I think `queryset.filter('functionName(x, y, z)')` can't allow to make something like
`queryset.filter(F('functionName', 'x', 'y', F('functionName2', 'x2', 'y2'))`, and support of "not only boolean functions" should be possible somehow through args (or additional kwargs)

another possible approach that maybe considered is to treat functions like fields with underscore or double underscore syntax for variables of that function, for example
`queryset.filter(pointInPolygon__lat__lon=...)`

**Comment by kalombos** [comment]
I want to share my workaround for the issue:

```python
class RawOperator(SimpleOperator):
    """
    Workaround for raw sql in conditions
    """

    def to_sql(self, model_cls, field_name, value):
        if field_name == 'raw':
            return '(%s)' % value
        field = getattr(model_cls, field_name)
        value = field.to_db_string(field.to_python(value, pytz.utc))
        return ' '.join([field_name, self._sql_operator, value])


register_operator('eq', RawOperator('='))

# How to use it
raw_condition = 'pointInPolygon((latitude, longitude), %s)' % coords
SomeModel.objects_in(db).filter(raw=raw_condition)
```

I think some feature like this also would be useful for cases when query api don't go far enough

**Comment by ishirav** [comment]
@kalombos - this is similar to the option I suggested above - `queryset.filter('functionName(x, y, z)')`
Your implementation requires adding the `raw=` before the condition string, but it's possible to omit it by changing the implementation of `filter()` to accept strings as raw conditions.
This seems to be the simplest solution, but maybe not the most helpful solution for the user. Basically we're telling him "the ORM can't help you with this, you need to use raw SQL".

**Comment by kalombos** [comment]
> This seems to be the simplest solution, but maybe not the most helpful solution for the user. Basically we're telling him "the ORM can't help you with this, you need to use raw SQL".

I am totally agree with you but the orm does not have any solution for the case at all. And i think there are another cases about which we don't know yet. I think opportunity to use raw sql should be in ORM.
Such opportunity exists in [Django ORM](https://docs.djangoproject.com/en/2.0/ref/models/expressions/#raw-sql-expressions) I don't propose to use raw sql as a solution for the issue, i propose to have an alternate universal future. Possibly i should open a new issue for the feature.

**Comment by M1hacka** [comment]
I've thought on this issue too, and it can be done like [django extra method](https://docs.djangoproject.com/en/2.1/ref/models/querysets/#django.db.models.query.QuerySet.extra). It will be helpful, as it can be used not only in filtering, but also can help integrate difficult SQL constructions to different parts of SQL in QuerySet without breaking QuerySet functions.
For instance, I want to have ability to use subqueries in from section (see https://github.com/Infinidat/infi.clickhouse_orm/issues/97)

**Comment by ishirav** [comment]
@M1hacka - I'm working on functions support, I think I have a good solution which will allow writing very expressive query filters... Once I finish I'll share it with you guys to get feedback.

**Comment by ishirav** [comment]
Available in v2.0.0
